File: swift/swiftsearch/Sources/swiftsearch/SearchOptions.swift
//
//  SearchOptions.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/17/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation
import swiftfind

struct SearchOption: Option {
    let shortArg: String?
    let longArg: String
    let desc: String
    let argType: ArgTokenType

    var sortArg: String {
        if shortArg != nil, !shortArg!.isEmpty {
            return shortArg!.lowercased() + "@" + longArg.lowercased()
        }
        return longArg.lowercased()
    }
}

public class SearchOptions {
    private var config: SearchConfig
    private var searchOptions = [SearchOption]()
    private var argTokenizer: ArgTokenizer?

    public init() {
        self.config = SearchConfig()
        setSearchOptionsFromJson()
        argTokenizer = ArgTokenizer(searchOptions)
    }

    private func setSearchOptionsFromJson() {
        do {
            let searchOptionsUrl = URL(fileURLWithPath: config.searchOptionsPath)
            let data = try Data(contentsOf: searchOptionsUrl, options: .mappedIfSafe)
            if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                if let options = json["searchoptions"] as? [[String: Any]] {
                    for so in options {
                        let longArg = so["long"] as! String
                        let shortArg = so.index(forKey: "short") != nil ? so["short"] as! String : ""
                        let desc = so["desc"] as! String
                        var argType = ArgTokenType.unknown
                        if self.boolActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.bool
                        } else if self.stringActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.str
                        } else if self.intActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.int
                        } else if self.longActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.long
                        }
                        searchOptions.append(SearchOption(shortArg: shortArg, longArg: longArg, desc: desc, argType: argType))
                    }
                }
            }
        } catch let error as NSError {
            print("Failed to load: \(error.localizedDescription)")
        }
    }

    private let boolActionDict: [String: (Bool, SearchSettings) -> Void] = [
        "allmatches": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.firstMatch = !bool
        },
        "archivesonly": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.archivesOnly = bool
        },
        "colorize": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.colorize = bool
        },
        "debug": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.debug = bool
        },
        "excludehidden": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.includeHidden = !bool
        },
        "firstmatch": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.firstMatch = bool
        },
        "followsymlinks": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.followSymlinks = bool
        },
        "help": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printUsage = bool
        },
        "includehidden": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.includeHidden = bool
        },
        "multilinesearch": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.multiLineSearch = bool
        },
        "nocolorize": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.colorize = !bool
        },
        "nofollowsymlinks": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.followSymlinks = !bool
        },
        "noprintdirs": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printDirs = !bool
        },
        "noprintfiles": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printFiles = !bool
        },
        "noprintlines": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printLines = !bool
        },
        "noprintmatches": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printResults = !bool
        },
        "norecursive": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.recursive = !bool
        },
        "nosearcharchives": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.searchArchives = !bool
        },
        "printdirs": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printDirs = bool
        },
        "printfiles": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printFiles = bool
        },
        "printlines": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printLines = bool
        },
        "printmatches": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printResults = bool
        },
        "recursive": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.recursive = bool
        },
        "searcharchives": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.searchArchives = bool
        },
        "sort-ascending": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.sortDescending = !bool
        },
        "sort-caseinsensitive": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.sortCaseInsensitive = bool
        },
        "sort-casesensitive": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.sortCaseInsensitive = !bool
        },
        "sort-descending": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.sortDescending = bool
        },
        "uniquelines": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.uniqueLines = bool
        },
        "verbose": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.verbose = bool
        },
        "version": { (bool: Bool, settings: SearchSettings) -> Void in
            settings.printVersion = bool
        }
    ]

    private var stringActionDict: [String: (String, SearchSettings) -> Void] = [
        "encoding": { (str: String, settings: SearchSettings) -> Void in
            settings.textFileEncoding = str
        },
        "in-archiveext": { (str: String, settings: SearchSettings) -> Void in
            settings.addInArchiveExtension(str)
        },
        "in-archivefilepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addInArchiveFilePattern(str)
        },
        "in-dirpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addInDirPattern(str)
        },
        "in-ext": { (str: String, settings: SearchSettings) -> Void in
            settings.addInExtension(str)
        },
        "in-filepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addInFilePattern(str)
        },
        "in-filetype": { (str: String, settings: SearchSettings) -> Void in
            settings.addInFileType(str)
        },
        "in-linesafterpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addInLinesAfterPattern(str)
        },
        "in-linesbeforepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addInLinesBeforePattern(str)
        },
        "linesaftertopattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addLinesAfterToPattern(str)
        },
        "linesafteruntilpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addLinesAfterUntilPattern(str)
        },
        "maxlastmod": { (str: String, settings: SearchSettings) -> Void in
            settings.setMaxLastModFromString(str)
        },
        "minlastmod": { (str: String, settings: SearchSettings) -> Void in
            settings.setMinSizeFromString(str)
        },
        "out-archiveext": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutArchiveExtension(str)
        },
        "out-archivefilepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutArchiveFilePattern(str)
        },
        "out-dirpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutDirPattern(str)
        },
        "out-ext": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutExtension(str)
        },
        "out-filepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutFilePattern(str)
        },
        "out-filetype": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutFileType(str)
        },
        "out-linesafterpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutLinesAfterPattern(str)
        },
        "out-linesbeforepattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addOutLinesBeforePattern(str)
        },
        "path": { (str: String, settings: SearchSettings) -> Void in
            settings.addPath(str)
        },
        "searchpattern": { (str: String, settings: SearchSettings) -> Void in
            settings.addSearchPattern(str)
        },
        "sort-by": { (str: String, settings: SearchSettings) -> Void in
            settings.setSortBy(str)
        }
    ]

    private let intActionDict: [String: (Int32, SearchSettings) -> Void] = [
        "linesafter": { (i: Int32, settings: SearchSettings) -> Void in
            settings.linesAfter = i
        },
        "linesbefore": { (i: Int32, settings: SearchSettings) -> Void in
            settings.linesBefore = i
        },
        "maxdepth": { (i: Int32, settings: SearchSettings) in
            settings.maxDepth = i
        },
        "maxlinelength": { (i: Int32, settings: SearchSettings) -> Void in
            settings.maxLineLength = i
        },
        "mindepth": { (i: Int32, settings: SearchSettings) in
            settings.minDepth = i
        }
    ]

    private let longActionDict: [String: (UInt64, SearchSettings) -> Void] = [
        "maxsize": { (l: UInt64, settings: SearchSettings) in
            settings.maxSize = l
        },
        "minsize": { (l: UInt64, settings: SearchSettings) in
            settings.minSize = l
        }
    ]

    public func updateSettingsFromArgTokens(_ settings: SearchSettings, argTokens: [ArgToken]) throws {
        for argToken in argTokens {
            if argToken.type == ArgTokenType.bool {
                if let bool = argToken.value as? Bool {
                    boolActionDict[argToken.name]!(bool, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else if argToken.type == ArgTokenType.str {
                if let string = argToken.value as? String {
                    if argToken.name == "settings-file" {
                        try updateSettingsFromFile(settings, filePath: string)
                    } else {
                        stringActionDict[argToken.name]!(string, settings)
                    }
                } else if let stringArray = argToken.value as? [String] {
                    for s in stringArray {
                        stringActionDict[argToken.name]!(s, settings)
                    }
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else if argToken.type == ArgTokenType.int {
                if let intVal = argToken.value as? Int32 {
                    intActionDict[argToken.name]!(intVal, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            }  else if argToken.type == ArgTokenType.long {
                if let longVal = argToken.value as? UInt64 {
                    longActionDict[argToken.name]!(longVal, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else {
                throw FindError(msg: "Invalid option: \(argToken.name)")
            }
        }
    }

    public func updateSettingsFromJson(_ settings: SearchSettings, jsonString: String) throws {
        let argTokens = try argTokenizer!.tokenizeJson(jsonString)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromJson(_ jsonString: String) throws -> SearchSettings {
        let settings = SearchSettings()
        try updateSettingsFromJson(settings, jsonString: jsonString)
        return settings
    }

    public func updateSettingsFromFile(_ settings: SearchSettings, filePath: String) throws {
        let argTokens = try argTokenizer!.tokenizeFile(filePath)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromFile(_ filePath: String) throws -> SearchSettings {
        let settings = SearchSettings()
        try updateSettingsFromFile(settings, filePath: filePath)
        return settings
    }

    public func updateSettingsFromArgs(_ settings: SearchSettings, args: [String]) throws {
        let argTokens = try argTokenizer!.tokenizeArgs(args)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromArgs(_ args: [String]) throws -> SearchSettings {
        let settings = SearchSettings()
        // default printResults to true since running in cli
        settings.printResults = true
        try updateSettingsFromArgs(settings, args: args)
        return settings
    }

    func getUsageString() -> String {
        var str = "\nUsage:\n swiftsearch [options] -s <searchpattern> <path> [<path> ...]\n\n"
        str += "Options:\n"
        let options = searchOptions.sorted(by: { $0.sortArg < $1.sortArg })

        let optStrings = options.map {
            switch ($0.shortArg, $0.longArg) {
            // Order errors by code
            case let (nil, longArg):
                "--\(longArg)"
            case let ("", longArg):
                "--\(longArg)"
            case let (shortArg, longArg):
                "-\(shortArg!),--\(longArg)"
            }
        }

        let optDescs = options.map(\.desc)
        let longest = optStrings.map { $0.lengthOfBytes(using: String.Encoding.utf8) }.max()!
        for i in 0 ..< optStrings.count {
            var optLine = " \(optStrings[i])"
            while optLine.lengthOfBytes(using: String.Encoding.utf8) <= longest {
                optLine += " "
            }
            optLine += "  \(optDescs[i])\n"
            str += optLine
        }
        return str
    }

    public func usage(_ code: Int32 = 0) {
        logMsg(getUsageString())
        exit(code)
    }
}

File: csharp/CsSearch/CsSearchLib/SearchOptions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using CsFindLib;
using SearchOptionsDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string,string>>>;

namespace CsSearchLib;

public class SearchOptions
{
	private readonly string _searchOptionsResource;

	private static readonly Dictionary<string, Action<bool, SearchSettings>> BoolActionDictionary =
		new()
		{
			{ "allmatches", (b, settings) => settings.FirstMatch = !b },
			{ "archivesonly", (b, settings) => settings.ArchivesOnly = b },
			{ "colorize", (b, settings) => settings.Colorize = b },
			{ "debug", (b, settings) => settings.Debug = b },
			{ "excludehidden", (b, settings) => settings.IncludeHidden = !b },
			{ "firstmatch", (b, settings) => settings.FirstMatch = b },
			{ "followsymlinks", (b, settings) => settings.FollowSymlinks = b },
			{ "help", (b, settings) => settings.PrintUsage = b },
			{ "includehidden", (b, settings) => settings.IncludeHidden = b },
			{ "multilinesearch", (b, settings) => settings.MultiLineSearch = b },
			{ "nocolorize", (b, settings) => settings.Colorize = !b },
			{ "nofollowsymlinks", (b, settings) => settings.FollowSymlinks = !b },
			{ "noprintdirs", (b, settings) => settings.PrintDirs = !b },
			{ "noprintfiles", (b, settings) => settings.PrintFiles = !b },
			{ "noprintlines", (b, settings) => settings.PrintLines = !b },
			{ "noprintmatches", (b, settings) => settings.PrintMatches = !b },
			{ "noprintresults", (b, settings) => settings.PrintResults = !b },
			{ "norecursive", (b, settings) => settings.Recursive = !b },
			{ "nosearcharchives", (b, settings) => settings.SearchArchives = !b },
			{ "printdirs", (b, settings) => settings.PrintDirs = b },
			{ "printfiles", (b, settings) => settings.PrintFiles = b },
			{ "printlines", (b, settings) => settings.PrintLines = b },
			{ "printmatches", (b, settings) => settings.PrintMatches = b },
			{ "printresults", (b, settings) => settings.PrintResults = b },
			{ "recursive", (b, settings) => settings.Recursive = b },
			{ "searcharchives", (b, settings) => settings.SearchArchives = b },
			{ "sort-ascending", (b, settings) => settings.SortDescending = !b },
			{ "sort-caseinsensitive", (b, settings) => settings.SortCaseInsensitive = b },
			{ "sort-casesensitive", (b, settings) => settings.SortCaseInsensitive = !b },
			{ "sort-descending", (b, settings) => settings.SortDescending = b },
			{ "uniquelines", (b, settings) => settings.UniqueLines = b },
			{ "verbose", (b, settings) => settings.Verbose = b },
			{ "version", (b, settings) => settings.PrintVersion = b },
		};

	private static readonly Dictionary<string, Action<string, SearchSettings>> StringActionDictionary =
		new()
		{
			{ "encoding", (s, settings) => settings.TextFileEncoding = s },
			{ "in-archiveext", (s, settings) => settings.AddInArchiveExtension(s) },
			{ "in-archivefilepattern", (s, settings) => settings.AddInArchiveFilePattern(s) },
			{ "in-dirpattern", (s, settings) => settings.AddInDirPattern(s) },
			{ "in-ext", (s, settings) => settings.AddInExtension(s) },
			{ "in-filepattern", (s, settings) => settings.AddInFilePattern(s) },
			{ "in-filetype", (s, settings) => settings.AddInFileType(s) },
			{ "in-linesafterpattern", (s, settings) => settings.AddInLinesAfterPattern(s) },
			{ "in-linesbeforepattern", (s, settings) => settings.AddInLinesBeforePattern(s) },
			{ "linesaftertopattern", (s, settings) => settings.AddLinesAfterToPattern(s) },
			{ "linesafteruntilpattern", (s, settings) => settings.AddLinesAfterUntilPattern(s) },
			{ "maxlastmod", (s, settings) => {
					settings.MaxLastMod = DateTime.Parse(s);
				}
			},
			{ "minlastmod", (s, settings) => {
					settings.MinLastMod = DateTime.Parse(s);
				}
			},
			{ "out-archiveext", (s, settings) => settings.AddOutArchiveExtension(s) },
			{ "out-archivefilepattern", (s, settings) => settings.AddOutArchiveFilePattern(s) },
			{ "out-dirpattern", (s, settings) => settings.AddOutDirPattern(s) },
			{ "out-ext", (s, settings) => settings.AddOutExtension(s) },
			{ "out-filepattern", (s, settings) => settings.AddOutFilePattern(s) },
			{ "out-filetype", (s, settings) => settings.AddOutFileType(s) },
			{ "out-linesafterpattern", (s, settings) => settings.AddOutLinesAfterPattern(s) },
			{ "out-linesbeforepattern", (s, settings) => settings.AddOutLinesBeforePattern(s) },
			{ "path", (s, settings) => settings.AddPath(s) },
			{ "searchpattern", (s, settings) => settings.AddSearchPattern(s) },
			{ "sort-by", (s, settings) => settings.SetSortBy(s) },
		};

	private static readonly Dictionary<string, Action<int, SearchSettings>> IntActionDictionary =
		new()
		{
			{ "linesafter", (i, settings) => settings.LinesAfter = i },
			{ "linesbefore", (i, settings) => settings.LinesBefore = i },
			{ "maxdepth", (i, settings) => settings.MaxDepth = i },
			{ "maxlinelength", (i, settings) => settings.MaxLineLength = i },
			{ "maxsize", (i, settings) => settings.MaxSize = i },
			{ "mindepth", (i, settings) => settings.MinDepth = i },
			{ "minsize", (i, settings) => settings.MinSize = i },
		};

	public List<IOption> Options { get; }
	private ArgTokenizer ArgTokenizer { get; }

	public SearchOptions()
	{
		_searchOptionsResource = EmbeddedResource.GetResourceFileContents("CsSearchLib.Resources.searchoptions.json");
		Options = [];
		SetOptionsFromJson();
		ArgTokenizer = new ArgTokenizer(Options);
	}

	private void SetOptionsFromJson()
	{
		var searchOptionsDict = JsonSerializer.Deserialize<SearchOptionsDictionary>(_searchOptionsResource);
		if (searchOptionsDict == null
		    || !searchOptionsDict.TryGetValue("searchoptions", out List<Dictionary<string, string>>? optionDicts))
		{
			throw new SearchException("Missing or invalid search options resource");
		}

        foreach (var optionDict in optionDicts)
		{
			var longArg = optionDict["long"];
			string? shortArg = null;
			if (optionDict.TryGetValue("short", out var shortVal))
			{
				shortArg = shortVal;
			}
			var desc = optionDict["desc"];
			ArgTokenType argType;
			if (BoolActionDictionary.ContainsKey(longArg))
			{
				argType = ArgTokenType.Bool;
			}
			else if (StringActionDictionary.ContainsKey(longArg) || longArg.Equals("settings-file"))
			{
				argType = ArgTokenType.String;
			}
			else if (IntActionDictionary.ContainsKey(longArg))
			{
				argType = ArgTokenType.Int;
			}
			else
			{
				throw new SearchException($"Invalid option: {longArg}");
			}
			var option = new SearchOption(shortArg, longArg, desc, argType);
			Options.Add(option);
		}
	}

	private void ApplyArgTokenToSettings(ArgToken argToken, SearchSettings settings)
	{
		if (argToken.Type == ArgTokenType.Bool)
		{
			if (BoolActionDictionary.TryGetValue(argToken.Name, out var boolAction))
			{
				if (argToken.Value is bool b)
				{
					boolAction(b, settings);
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.False or JsonValueKind.True } jsonElem)
				{
					boolAction(jsonElem.GetBoolean(), settings);
				}
				else
				{
					throw new SearchException($"Invalid value for option: {argToken.Name}");
				}
			}
			else
			{
				throw new SearchException($"Invalid option: {argToken.Name}");
			}
		}
		else if (argToken.Type == ArgTokenType.String)
		{
			if (StringActionDictionary.TryGetValue(argToken.Name, out var stringAction))
			{
				if (argToken.Value is string s)
				{
					stringAction(s, settings);
				}
				else if (argToken.Value is IEnumerable<string> enumerable)
				{
					foreach (var enumVal in enumerable)
					{
						stringAction(enumVal, settings);
					}
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.String }  jsonElem)
				{
					var jsonString = jsonElem.GetString();
					if (jsonString is not null)
					{
						stringAction(jsonString, settings);
					}
					else
					{
						throw new SearchException($"Invalid value for option: {argToken.Name}");
					}
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.Array }  jsonArr)
				{
					foreach (var arrVal in jsonArr.EnumerateArray())
					{
						if (arrVal is JsonElement { ValueKind: JsonValueKind.String }  jsonArrElem)
						{
							var jsonStr = jsonArrElem.GetString();
							if (jsonStr is not null)
							{
								stringAction(jsonStr, settings);
							}
							else
							{
								throw new SearchException($"Invalid value for option: {argToken.Name}");
							}
						}
					}
				}
				else
				{
					throw new SearchException($"Invalid value for option: {argToken.Name}");
				}
			}
			else if (argToken.Name.Equals("settings-file"))
			{
				if (argToken.Value is string settingsFilePath)
				{
					UpdateSettingsFromFile(settings, settingsFilePath);
				}
				else if (argToken.Value is IEnumerable<string> enumerable)
				{
					foreach (var enumVal in enumerable)
					{
						UpdateSettingsFromFile(settings, enumVal);
					}
				}
				else
				{
					throw new SearchException($"Invalid value for option: {argToken.Name}");
				}
			}
			else
			{
				throw new SearchException($"Invalid option: {argToken.Name}");
			}
		}
		else if (argToken.Type == ArgTokenType.Int)
		{
			if (IntActionDictionary.TryGetValue(argToken.Name, out var intAction))
			{
				if (argToken.Value is int i)
				{
					intAction(i, settings);
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.Number } jsonElem)
				{
					intAction(jsonElem.GetInt32(), settings);
				}
				else
				{
					throw new SearchException($"Invalid value for option: {argToken.Name}");
				}
			}
		}
		else
		{
			throw new SearchException($"Invalid option: {argToken.Name}");
		}
	}

	private void UpdateSettingsFromArgTokens(SearchSettings settings, List<ArgToken> argTokens)
	{
		foreach (var argToken in argTokens)
		{
			ApplyArgTokenToSettings(argToken, settings);
		}
	}

	private void UpdateSettingsFromDictionary(SearchSettings settings, Dictionary<string, object> dictionary)
	{
		var argTokens = ArgTokenizer.TokenizeDictionary(dictionary);
		UpdateSettingsFromArgTokens(settings, argTokens);
	}

	public void UpdateSettingsFromJson(SearchSettings settings, string jsonString)
	{
		try
		{
			var argTokens = ArgTokenizer.TokenizeJson(jsonString);
			UpdateSettingsFromArgTokens(settings, argTokens);
		} catch (FindException e)
		{
			throw new SearchException(e.Message);
		}
	}

	public SearchSettings SettingsFromJson(string jsonString)
	{
		var settings = new SearchSettings();
		UpdateSettingsFromJson(settings, jsonString);
		return settings;
	}

	public void UpdateSettingsFromFile(SearchSettings settings, string filePath)
	{
		try
		{
			var argTokens = ArgTokenizer.TokenizeFile(filePath);
			UpdateSettingsFromArgTokens(settings, argTokens);
		} catch  (FindException e)
		{
			throw new SearchException(e.Message);
		}
	}

	public SearchSettings SettingsFromFile(string filePath)
	{
		var settings = new SearchSettings();
		UpdateSettingsFromFile(settings, filePath);
		return settings;
	}

	public void UpdateSettingsFromArgs(SearchSettings settings, IEnumerable<string> args)
	{
		try
		{
			var argTokens = ArgTokenizer.TokenizeArgs(args);
			UpdateSettingsFromArgTokens(settings, argTokens);
		}
		catch (FindException e)
		{
			throw new SearchException(e.Message);
		}
	}

	public SearchSettings SettingsFromArgs(IEnumerable<string> args)
	{
		// default to PrintResults = true since this is called from CLI
		var settings = new SearchSettings {PrintResults = true};
		UpdateSettingsFromArgs(settings, args);
		return settings;
	}

	public string GetUsageString()
	{
		var sb = new StringBuilder();
		sb.AppendLine("\nUsage:");
		sb.AppendLine(" cssearch [options] -s <searchpattern> <path> [<path> ...]\n");
		sb.AppendLine("Options:");
		var optStrings = new List<string>();
		var optDescs = new List<string>();
		var longest = 0;
		foreach (var opt in Options.OrderBy(o => o.SortArg))
		{
			var optString = new StringBuilder();
			if (!string.IsNullOrWhiteSpace(opt.ShortArg))
			{
				optString.Append("-" + opt.ShortArg + ",");
			}
			optString.Append("--" + opt.LongArg);
			if (optString.Length > longest)
			{
				longest = optString.Length;
			}
			optStrings.Add(optString.ToString());
			optDescs.Add(opt.Description);
		}
		var format = " {0,-"+longest+"}  {1}";
		for (var i = 0; i < optStrings.Count; i++)
		{
			sb.AppendLine(string.Format(format, optStrings[i], optDescs[i]));
		}
		return sb.ToString();
	}

	public void Usage(int exitCode = 0)
	{
		Console.WriteLine(GetUsageString());
		Environment.Exit(exitCode);
	}
}
